src/dbnode/storage/namespace_readers.go

```diff
@@ -380,3 +444,26 @@ func (m *namespaceReaderManager) tickWithThreshold(threshold int) {
 		m.openReaders[key] = elem
 	}
 }
+
+// UpdateOpenLease() implements block.Leaser.
+func (m *namespaceReaderManager) UpdateOpenLease(
+	descriptor block.LeaseDescriptor,
+	state block.LeaseState,
+) (block.UpdateOpenLeaseResult, error) {
+	if !m.namespace.ID().Equal(descriptor.Namespace) {
+		return block.NoOpenLease, nil
+	}
+
+	m.Lock()
+	// Remove open readers with matching key but lower volume.
+	for readerKey := range m.openReaders {
+		if readerKey.shard == descriptor.Shard &&
+			readerKey.blockStart == xtime.ToUnixNano(descriptor.BlockStart) &&
+			readerKey.volume < state.Volume {
+			delete(m.openReaders, readerKey)
```

Comment:
dont you need to actually close it?

Reply:
Hah, good point. I'll also put it back to the `closedReaders` which serves as the pool of readers here.